public/js/helpers/zoom_helper.js

```diff
@@ -112,16 +112,31 @@ export default class Zoom {
 
   // mapKey returns the corresponding map key, if the zoom meets the correct
   // range and position within the limits, else null.
-  static mapKey (zoom, limits, scale) {
+  static mapKey (zoom, limits, scale, isLegacyURL) {
     scale = scale || 1
+    isLegacyURL = isLegacyURL || false
     let lims = tryDecode(limits)
     let decoded = this.decode(zoom, lims)
-    let range = decoded.end - decoded.start
-    for (let k in zoomMap) {
-      let v = zoomMap[k]
-      if (v / scale === range) return k
+
+    if (isLegacyURL) { // maps the legacy url zoom lookup
+      let range = decoded.end - decoded.start
+      for (let k in zoomMap) {
+        let v = zoomMap[k]
+        if (v / scale === range) return k
+      }
+      return 'all'
```

Comment:
![image](https://user-images.githubusercontent.com/6109680/58504564-8299e000-8150-11e9-8a03-9f07eccae7d0.png)

I am still completely uncertain about what you are trying to accomplish by changing this function. The output of this function is only used to set what button is blue. It does not affect any URL query parameters. It only sets the button color. Was there an issue with the way the buttons were behaving? What did you see that made you want to change this? 

To make this work correctly with `setActiveOptionBtn`, you should return this function to its original form but make sure you pass a numerical `scale` (or add your `scale = scale || 1` in `mapKey`). Remove this line https://github.com/dmigwi/dcrdata/blob/0a5b19c78d91cf325dd3691e39b423c0bbfeaeaa/public/js/controllers/charts_controller.js#L473

and add this line

```
this.setActiveOptionBtn(Zoom.mapKey(this.settings.zoom, this.limits), this.zoomOptionTargets)
```

at the end of `_zoomCallback` and `_drawCallback`.

Reply:
The zoom on this url http://localhost:7777/charts?chart=ticket-pool-value&zoom=juy57sw0-jw5luyo0&bin=day&axis=time corresponds to `month` zoom interval. 
Without my `mapKey` implementation the zoom in that url doesn't show the month zoomed chart.

Reply:
I'm willing to bet it's because you were passing a bad value for the `scale` argument at some point. I've tested with the changes I recommended above, and it is working correctly.

Reply:
Removing this line https://github.com/dmigwi/dcrdata/blob/0a5b19c78d91cf325dd3691e39b423c0bbfeaeaa/public/js/controllers/charts_controller.js#L473 will mean that  when a chart corresponding to `all` zoom type is visible then the `all` button may or may not active.

I think the effect you need functioning should be fixing by just adding:
```
this.setActiveOptionBtn(Zoom.mapKey(this.settings.zoom, this.limits), this.zoomOptionTargets)
```
to the 
`_zoomCallBack`.

Reply:
All should be well now.

Reply:
That will not fix the problem in the image I pasted above. The function `mapKey` function was working perfectly before. There was no need to change it.

Reply:
Removing that line will not make it work incorrectly.

Reply:
The same issue I was trying to fix exists on alpha: https://alpha.dcrdata.org/charts?chart=ticket-price&zoom=juzfs5nk-jw6wfbfk. This zoom is set to the month, but the month button isn't activated or page hard reloading.

Try it on your end using alpha:
1. Select any zoom button of your choice and copy the chart URL.
2. Open a brand new tab and load the chart URL you just copied. Let me know if the specific zoom button you selected earlier is activated on the new tab.

This is what my `mapKey` implementation is supposed to fix.

Reply:
I think I might see what you're saying. So even though the URL ` https://alpha.dcrdata.org/charts?chart=ticket-price&zoom=month` will set the range and selected button correctly, an encoded zoom range like `zoom=juzfs5nk-jw6wfbfk`, which should correspond to a month, only sets the range correctly but not the selected button. 

Does that sound right?

Reply:
Thats it.

Reply:
So there are a couple of subtleties here. 

When you give somebody a URL for the charts page, it should always bring them to exactly the view that's on the chart when you copied the URL (which is why the keyword is immediately replaced with an encoded range). As mentioned before, the keywords correspond to only the most recent window. So if the *range* is say a month, but the end of the range is not at the rightmost point of the available data, the button is not shown as selected. That means that an encoded-range URL that corresponds to a valid button window will only set the button selection if used before the next block is found, at which point it's rightmost point is no longer at the extent of the available data.

It is true that even if you immediately re-paste the URL with encoded range it does not set the selected button correctly. It's certainly reasonable to want to fix that, though again, it only matters til the next block is found anyway. It's also reasonable to want show the button as selected based only on range and not on whether it is the most recent possible window of that duration. Neither of those solutions requires any new parameters or special case handling in the `mapKey` function.

How about you use your range matching logic, but with a couple tweaks as below. 

1. Default value needs to be `null`. 
1. Special check for `all`
```
  static mapKey (zoom, limits, scale) {
    scale = scale || 1
    let lims = tryDecode(limits)
    let decoded = this.decode(zoom, lims)
    let range = decoded.end - decoded.start
    if (range === lims.end - lims.start) return 'all'
    for (let k in zoomMap) {
      let v = zoomMap[k]
      if (v / scale === range) return k
    }
    return null
  }
```

You will need to remove the `if (!option) return` line for it to work correctly since `null` is a valid value. 

If after those changes, it still doesn't select the button correctly, even if you paste the URL before the next block, then it will probably just be a math error somewhere and we can chase that down.

Reply:
This should work, I will be testing that mapKey implementation in a few. Its looks just like the compromise we set out to achieve.

Reply:
I do like this change. It'll be a nice indicator that the window hasn't changed size.